node/pkg/txverifier/utils.go

```diff
@@ -16,13 +16,30 @@ import (
 const (
 	MAX_DECIMALS = 8
 	KEY_FORMAT   = "%s-%d"
+
+	// Invariant violation messages. There technically only exists two violations.
+	INVARIANT_NO_DEPOSIT           = "bridge transfer requested for tokens that were never deposited"
+	INVARIANT_INSUFFICIENT_DEPOSIT = "bridge transfer requested for more tokens than were deposited"
 )
 
+// Custom error type used to signal that a core invariant of the token bridge has been violated.
+type InvariantError struct {
+	Msg string
+}
+
+func (i InvariantError) Error() string {
+	return fmt.Sprintf("invariant violated: %s", i.Msg)
+}
+
 // Extracts the value at the given path from the JSON object, and casts it to
 // type T. If the path does not exist in the object, an error is returned.
 func extractFromJsonPath[T any](data json.RawMessage, path string) (T, error) {
 	var defaultT T
 
+	if data == nil {
```

Comment:
```suggestion
	if len(data) == 0 {
```
Can `json.RawMessage` (alias for `[]byte`) be `nil` here? In any case the length check will also cover the `nil` case.

Reply:
My hunch is that this should return an error if the JSON is empty, regardless of the nil check. We could do both if you want to be extra clear about it.